rfd/0197-prometheus-metrics.md

```diff
@@ -0,0 +1,237 @@
+---
+authors: Hugo Hervieux (hugo.hervieux@goteleport.com)
+state: draft
+---
+
+# RFD 197 - Prometheus metrics guidelines
+
+## Required Approvers
+
+* Engineering: @codingllama && @rosstimothy && @zmb3
+
+## What
+
+This RFD covers the recommended way of adding metrics to Teleport and other services such as
+tbot, the Teleport Kube Agent Updater, the Teleport Kubernetes Operator, or plugins
+(slack, teams, pagerduty, event-handler, ...).
+
+## Why
+
+Every component currently relies on the global metrics registry and package-scoped metrics.
+This poses several challenges:
+- We pick up and serve arbitrary metrics declared by any of our dependencies. This can cause registration conflicts
+  (errors, flakiness, panics), high cardinality, confusing metric names, and data loss.
+- Several Teleport components cannot run in the same go process without causing metric conflicts. This is the case in
+  integration tests (running both tbot and Teleport), or even in user-facing binaries (tctl uses `embeddedtbot`).
+- The same Teleport component cannot be started multiple times without causing metric conflicts, or inaccurate metrics.
+  For example, integration tests starting several Teleport instances will merge their metrics together.
+
+## Guidelines
+
+### Metrics Registry
+
+#### Do
+
+- Take the local in-process registry as an argument in your service constructor/main routine, like you would receive a
+  logger, and register your metrics against it.
+- Pass the registry as a `promtetheus.Registerer`
+
+```golang
+func NewService(log *slog.Logger, reg prometheus.Registerer) (Service, error) {
+    myMetric := prometheus.NewGauge(prometheus.GaugeOpts{
+        Namespace: teleport.MetricNamespace,
+        Subsystem: metricsSubsystem,
+        Name: "my_metric",
+        Help: "Measures the number of foos doing bars.",
+    }),
+    if err := reg.Register(myMetric); err != nil {
+        return nil, trace.Wrap(err, "registering metric")
+    }
+    // ...
+}
+```
+
+#### Don't
+
+- Register against the global prometheus registry
```

Comment:
Suggestion: I think "Don't" sections could be mainly bullets, so readers focus more on the "Do"s (and have less risk of copying the Don't code).